store/sqlstore/migration.go

```diff
@@ -0,0 +1,194 @@
+func (r *MigrationRunner) Add(m AsyncMigration) error {
+	// check status in Systems table
+	currentStatus, appErr := r.supplier.System().GetByName("migration_" + m.Name())
+	if appErr == nil {
+		if currentStatus.Value == string(AsyncMigrationStatusComplete) || currentStatus.Value == string(AsyncMigrationStatusSkip) {
+			return nil
+		}
+	}
+	// get status from migration
+	status, err := m.GetStatus(r.supplier)
+	if err != nil {
+		return err
+	}
+	if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+		return nil
+	}
+	r.migrations = append(r.migrations, m)
+	return nil
+}
+
+// Run all queued migrations sequentially
+func (r *MigrationRunner) Run() error {
+	go func() {
+		ctx, cancelFn := context.WithCancel(context.Background())
+		defer cancelFn()
+		defer close(r.done)
+		r.cancelFn = cancelFn
+
+		for _, m := range r.migrations {
+			// function that will try to execute migration
+			migrate := func() error {
+				var err error
+				conn, err := createConnectionWithLockTimeout(ctx, r.supplier, r.options.LockTimeout)
+				if err != nil {
+					mlog.Error("Failed to setup connection", mlog.Err(err))
+					return err
+				}
+				defer releaseConnection(ctx, r.supplier, conn)
+
+				// run migration
+				status, err := m.Execute(ctx, r.supplier, conn)
+				if err != nil {
+					mlog.Error("Failed to execute migration", mlog.Err(err))
+					return err
+				}
+				if status == AsyncMigrationStatusComplete || status == AsyncMigrationStatusSkip {
+					// save migration status
+					r.supplier.System().SaveOrUpdate(&model.System{Name: "migration_" + m.Name(), Value: string(status)})
+				} else if status == AsyncMigrationStatusFailed {
+					mlog.Error("Failed migration " + m.Name())
+					return errors.New("Failed migration")
+				} else {
+					// should we return error here to retry?
+					return errors.New("Invalid result from migration")
+				}
+				return nil
+			}
+			// retry migration if it fails
+			for i := 0; i < r.options.NumRetries; i++ {
+				err := migrate()
+				if err == nil {
+					break
+				}
+				select {
+				case <-ctx.Done():
+					return
+				// wait before trying again
+				case <-time.After(time.Duration(r.options.BackoffTime) * time.Second):
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// Wait returns after all migrations are processed (executed, skipped or failed)
+func (r *MigrationRunner) Wait() {
+	<-r.done
+}
+
+// Cancel running migrations
+func (r *MigrationRunner) Cancel() {
+	if r.cancelFn != nil {
+		r.cancelFn()
+	}
+}
+
+// get an explicit connection because that guarantees a single session for all queries
+func createConnectionWithLockTimeout(ctx context.Context, s SqlStore, timeout int) (*sql.Conn, error) {
+	conn, err := s.GetMaster().Db.Conn(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("SET SESSION lock_timeout = '%ds'", timeout))
```

Comment:
Let's refactor out the executing part from the if-else, and just decide the statement to run. And then have a single place to do `conn.ExecContext` and err-check.